### ISSUE TITLE
FEATURE: Mock Client() for local testing

### ISSUE THREAD

**Issue by nicklausbrown** [v3io/v3io-py#55]
To make testing code which relies on the v3io client simpler and implementation code more consistent, I believe this library should ship with a mock Client that implements the API but returns structured logging rather than actual v3io write/reads. The structured logs can be parsed by test suites to determine accurate behavior.

```
from v3io.dataplane import MockClient
client = MockClient()
client.kv.put(**kwargs)
# logging __main__.py: {“message”: “V3io key value put record”, “key”: “some key”}
```

**Comment by pavius** [comment]
Thanks for the input. Would like to propose an alternative to fill the need. The `Client` is decoupled from the `Transport` which does the actual v3io writes/reads. The well defined transport interface has two implementations: a "low level, high performance" Python HTTP client and the requests library, which is there as a backup.

I propose allowing the `Client()` caller to provide a custom transport class. This will allow tests not only to catch and validate requests through standard Python mocking libraries (request body will be as a dict, not as log so it'll be easier to access) but also craft responses to inject errors and such. 

WDYT?

**Comment by ihs-nick** [comment]
I think a custom transport class would be a great idea to simplify my original design idea. I think an example of how to write a custom transport class for testing would be good to have in the README documentation - after all, most people using this framework are data scientists not software engineers. The other option is to include a basic testing transport type as part of the library too with the dict response. Thoughts?

**Comment by pavius** [comment]
Sounds good.

I will:
1. Allow users to provide custom transports to Client()
2. Include a transport that will do basic request validation / response injection
3. Add a blurb to the README that will describe how to use the transport in (2) and how to implement custom transports for advanced mocking

**Comment by ihs-nick** [comment]
Great. I look forward to the implementation!

**Comment by pavius** [comment]
I released 0.5.4 with a verifying transport and support for custom transports. The idea is that the request verification/response generation is done before calling the actual code (similar to how mocking libraries do this). A usage example can be found in the README along with a blurb about custom transports:

https://github.com/v3io/v3io-py#support-for-unit-tests-experimental

Note that I opted for callables rather than doing the verification because callables allow the user much more flexibility at the expense of verboseness. As time goes by and use cases for the verifier find some common ground - this can be shoved into the verifier.
